File: C/newton_mod.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef struct
{
    long double x;
    long double y;
    long double z;
}Punto;

long double f1(long double x, long double y, long double z)
{
    return (x * x) + y + (2.0 * (z * z)) - 10.0;
}

long double f2(long double x, long double y, long double z)
{
    return (5 * x) - (6 * y) + z;
}

long double f3(long double x, long double y, long double z)
{
    return z - (x * x) - (y * y);
}
//con respecto a x
long double f1d(long double x, long double y, long double z)
{
    return 2.0 * x;
}
//con respecto a y
long double f2d(long double x, long double y, long double z)
{
    return -6.0;
}
//con respecto a z
long double f3d(long double x, long double y, long double z)
{
    return 1.0;
}

long double norma(Punto A)
{
    long double s = 0.0;

    s += (A.x * A.x);
    s += (A.y * A.y);
    s += (A.z * A.z);

    return sqrt(s);
}

Punto newton_mod(Punto A, long double tolerancia, int iteraciones)
{
    Punto X = A;
    Punto F_X_k, dF_X_k;
    long double e_a;
    int k = 0;

    while(1)
    {

        F_X_k.x = f1(X.x, X.y, X.z);
        F_X_k.y = f2(X.x, X.y, X.z);
        F_X_k.z = f3(X.x, X.y, X.z);

        dF_X_k.x = f1d(X.x, X.y, X.z);
        dF_X_k.y = f2d(X.x, X.y, X.z);
        dF_X_k.z = f3d(X.x, X.y, X.z);

        e_a = norma(F_X_k);

        printf("|%-4d|%-11.10LF|%11.10LF|%11.10LF|%-10.10LF|\n", k, X.x, F_X_k.x, dF_X_k.x, e_a);
        printf("|    |%-11.10LF|%11.10LF|%11.10LF|\n", X.y, F_X_k.y, dF_X_k.y);
        printf("|    |%-11.10LF|%11.10LF|%11.10LF|\n\n", X.z, F_X_k.z, dF_X_k.z);

        if (e_a <= tolerancia) break;
        if (k >= iteraciones) break;

        X.x = X.x - ((F_X_k.x)/(dF_X_k.x));
        X.y = X.y - ((F_X_k.y)/(dF_X_k.y));
        X.z = X.z - ((F_X_k.z)/(dF_X_k.z));

        k++;
    }

    return X;
}

int main()
{
    Punto X, R;

    X.x = 1.0;
    X.y = 1.0;
    X.z = 2.0;

    R = newton_mod(X, 0.5, 10);

    return 0;
}

File: Sistemas de ecuaciones no lineales/Newton/Julia/newton_mod.jl
#using LinearAlgebra
using Printf

function f1(x::Float64, y::Float64, z::Float64)
    return (x ^ 2) + y + (2 * (z ^ 2)) - 10.0
end

function f2(x::Float64, y::Float64, z::Float64)
    return (5 * x) - (6 * y) + z
end

function f3(x::Float64, y::Float64, z::Float64)
    return z - (x ^ 2) - (y ^ 2)
end
#con respecto a x
function f1d(x::Float64, y::Float64, z::Float64)
    return 2 * x
end
#con respecto a y
function f2d(x::Float64, y::Float64, z::Float64)
    return -6.0
end
#con respecto a z
function f3d(x::Float64, y::Float64, z::Float64)
    return 1.0
end

#En la libería estándar existe en LinearAlgebra la función norma que hace lo mismo. Meh.
function norma(A::Array{Float64})
    ∑ = 0.0

    for i in 1:length(A)
        ∑ += (A[i] ^ 2)
    end

    return √∑

end

#Usando despazamientos simultaneos
function newton_mod(X::Array{Float64}, tolerancia::Float64, iteraciones::Int64)

    X_km1 = X
    X_k = X
    F_X_k_ = [0.0, 0.0, 0.0]
    δF_X_k_ = [0.0, 0.0, 0.0]
    e_a = 1.0
    k = 0

    print("X = ")
    println(X_km1)

    while true

        X_km1 = X_k

        F_X_k_[1] = f1(X_km1[1], X_km1[2], X_km1[3])
        F_X_k_[2] = f2(X_km1[1], X_km1[2], X_km1[3])
        F_X_k_[3] = f3(X_km1[1], X_km1[2], X_km1[3])

        print("F(X) = ")
        println(F_X_k_)

        δF_X_k_[1] = f1d(X_km1[1], X_km1[2], X_km1[3])
        δF_X_k_[2] = f2d(X_km1[1], X_km1[2], X_km1[3])
        δF_X_k_[3] = f3d(X_km1[1], X_km1[2], X_km1[3])

        print("δF(X) = ")
        println(δF_X_k_)

        e_a = norma(F_X_k_)

        println(e_a)

        if  e_a <= tolerancia
            break
        end
        if k >= iteraciones
            break
        end

        X_k[1] = X_km1[1] - ((F_X_k_[1])/(δF_X_k_[1]))
        X_k[2] = X_km1[2] - ((F_X_k_[2])/(δF_X_k_[2]))
        X_k[3] = X_km1[3] - ((F_X_k_[3])/(δF_X_k_[3]))

        print("K = ")
        println(k)
        print("X^k = ")
        println(X_k)

        k += 1
    end

    return X_k

end
#=
function main()
    A = [0.5, 0.5]
    println(A)
    R = newton_mod(A, 0.00005, 10)
    println(R)
end

main()
=#
